lib/evaluate/host.h

```diff
@@ -0,0 +1,229 @@
+namespace Fortran::evaluate {
+namespace host {
+
+// Type mapping from F18 types to host types
+struct UnsupportedType {};  // There is no host type for the F18 type
+
+template<typename FTN_T> struct HostTypeHelper {
+  using Type = UnsupportedType;
+};
+template<typename FTN_T> using HostType = typename HostTypeHelper<FTN_T>::Type;
+
+template<typename... T> constexpr inline bool HostTypeExists() {
+  return (... && (!std::is_same_v<HostType<T>, UnsupportedType>));
+}
+
+// Type mapping from host types to F18 types FortranType<HOST_T> is defined
+// after all HosTypeHelper definition because it reverses them to avoid
+// duplication.
+
+// Scalar conversion utilities from host scalars to F18 scalars
+template<typename FTN_T>
+inline constexpr Scalar<FTN_T> CastHostToFortran(const HostType<FTN_T> &x) {
+  static_assert(HostTypeExists<FTN_T>());
+  if constexpr (FTN_T::category == TypeCategory::Complex &&
+      sizeof(Scalar<FTN_T>) != sizeof(HostType<FTN_T>)) {
+    // X87 is usually padded to 12 or 16bytes. Need to cast piecewise for
+    // complex
+    return Scalar<FTN_T>{CastHostToFortran<typename FTN_T::Part>(std::real(x)),
+        CastHostToFortran<typename FTN_T::Part>(std::imag(x))};
+  } else {
+    return *reinterpret_cast<const Scalar<FTN_T> *>(&x);
+  }
+}
+
+// Scalar conversion utilities from  F18 scalars to host scalars
+template<typename FTN_T>
+inline constexpr HostType<FTN_T> CastFortranToHost(const Scalar<FTN_T> &x) {
+  static_assert(HostTypeExists<FTN_T>());
+  if constexpr (FTN_T::category == TypeCategory::Complex &&
+      sizeof(Scalar<FTN_T>) != sizeof(HostType<FTN_T>)) {
+    // X87 is usually padded to 12 or 16bytes. Need to cast piecewise for
+    // complex
+    return HostType<FTN_T>{CastFortranToHost<typename FTN_T::Part>(x.REAL()),
+        CastFortranToHost<typename FTN_T::Part>(x.AIMAG())};
+  } else {
+    return *reinterpret_cast<const HostType<FTN_T> *>(&x);
+  }
+}
+
+template<typename T> struct BiggerOrSameHostTypeHelper {
+  using Type =
+      std::conditional_t<HostTypeExists<T>(), HostType<T>, UnsupportedType>;
+  using FortranType = T;
+};
+
+template<typename FTN_T>
+using BiggerOrSameHostType = typename BiggerOrSameHostTypeHelper<FTN_T>::Type;
+template<typename FTN_T>
+using BiggerOrSameFortranTypeSupportedOnHost =
+    typename BiggerOrSameHostTypeHelper<FTN_T>::FortranType;
+
+template<typename... T> constexpr inline bool BiggerOrSameHostTypeExists() {
+  return (... && (!std::is_same_v<BiggerOrSameHostType<T>, UnsupportedType>));
+}
+
+// Defining the actual mapping
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 1>> {
+  using Type = std::int8_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 2>> {
+  using Type = std::int16_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 4>> {
+  using Type = std::int32_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 8>> {
+  using Type = std::int64_t;
+};
+
+// no int 128bit
```

Comment:
Why not?

Reply:
I do not know what I could try mapping here. I think std::int128_t does no exit. I can maybe try to add a conditional type on "long long int" if it is 128bit. Do you know of any architecture where this is the case?

Reply:
Both GCC and clang have `__int128_t`.

Reply:
OK, I will add `__int128_t` with some ifdef, it seems that it is not defined on every architecture  ([src](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fint128.html)) so I will also check that `__SIZEOF_INT128__` is defined ([src](https://gcc.gnu.org/ml/gcc-help/2015-08/msg00185.html)) and protect the definition with:
`#if (defined(__GNUC__) || defined(__clang__)) && defined(__SIZEOF_INT128__)`
Also, I am confused regarding what is the difference between  `__int128` and `__int128_t` but I get a warning from g++ when using `__int128` that I do not get with `__int128_t` so I will go for this last one.

> ISO C++ does not support ‘__int128’ for ‘type name’ [-Wpedantic]